File: packages/ms-mde/src/policy.ts
import * as v from "valibot";
import { envelope, withEnvelope, xsiNil } from "./envelope";

// Contains device cryptographic information and user authentication information
export const policyRequest = envelope(
	"http://schemas.microsoft.com/windows/pki/2009/01/enrollmentpolicy/IPolicy/GetPolicies",
	{
		"@_xmlns:xsi": v.literal("http://www.w3.org/2001/XMLSchema-instance"),
		"@_xmlns:xsd": v.literal("http://www.w3.org/2001/XMLSchema"),
		GetPolicies: v.object({
			"@_xmlns": v.literal(
				"http://schemas.microsoft.com/windows/pki/2009/01/enrollmentpolicy",
			),
			client: v.object({
				lastUpdate: v.object({
					"@_xsi:nil": v.literal("true"),
				}),
				preferredLanguage: v.object({
					"@_xsi:nil": v.literal("true"),
				}),
				TPMManufacturer: v.string(),
				TPMFirmwareVersion: v.string(),
			}),
			requestFilter: v.object({
				"@_xsi:nil": v.literal("true"),
			}),
		}),
	},
	{
		schemas: {
			u: "http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd",
			wsse: "http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd",
			wst: "http://docs.oasis-open.org/ws-sx/ws-trust/200512",
			ac: "http://schemas.xmlsoap.org/ws/2006/12/authorization",
		},
	},
);

type PolicyResponseArgs = {
	policyId: string;
	policyFriendlyName: string;
};

// The policy defining how the device identity certificate will be generated
export const policyResponse = withEnvelope(
	"http://schemas.microsoft.com/windows/pki/2009/01/enrollmentpolicy/IPolicy/GetPoliciesResponse",
	(opts: PolicyResponseArgs) => ({
		"@_xmlns:xsi": "http://www.w3.org/2001/XMLSchema-instance",
		GetPoliciesResponse: {
			"@_xmlns":
				"http://schemas.microsoft.com/windows/pki/2009/01/enrollmentpolicy",
			response: {
				policyID: opts.policyId,
				policyFriendlyName: opts.policyFriendlyName,
				nextUpdateHours: xsiNil,
				policiesNotChanged: xsiNil,
				policies: {
					policy: {
						policyOIDReference: 0, // References to OID defined in OIDs section
						cAs: xsiNil,
						attributes: {
							policySchema: 3,
							privateKeyAttributes: {
								minimalKeyLength: 4096,
								keySpec: xsiNil,
								keyUsageProperty: xsiNil,
								permissions: xsiNil,
								algorithmOIDReference: xsiNil,
								cryptoProviders: xsiNil,
							},
							supersededPolicies: xsiNil,
							privateKeyFlags: xsiNil,
							subjectNameFlags: xsiNil,
							enrollmentFlags: xsiNil,
							generalFlags: xsiNil,
							hashAlgorithmOIDReference: 0,
							rARequirements: xsiNil,
							keyArchivalAttributes: xsiNil,
							extensions: xsiNil,
						},
					},
				},
			},
			cAs: "",
			oIDs: {
				policyOIDReference: 0,
				defaultName: "szOID_OIWSEC_SHA256",
				group: 2, // 2 = Encryption algorithm identifier
				value: "2.16.840.1.101.3.4.2.1",
			},
		},
	}),
);

File: crates/ms-mde/src/policy.rs
// const policyActionRequest = "http://schemas.microsoft.com/windows/pki/2009/01/enrollmentpolicy/IPolicy/GetPolicies"
// const policyActionResponse = "http://schemas.microsoft.com/windows/pki/2009/01/enrollmentpolicy/IPolicy/GetPoliciesResponse"

// package soap

// import (
// 	"github.com/mattrax/xml"
// )

// // PolicyRequest contains the user authentication information
// type PolicyRequest struct {
// 	XMLName xml.Name      `xml:"s:Envelope"`
// 	Header  RequestHeader `xml:"s:Header"`
// }

// // PolicyResponse contains the policy defining how the identity certificate and key will be generated
// type PolicyResponse struct {
// 	XMLName  xml.Name           `xml:"http://schemas.microsoft.com/windows/pki/2009/01/enrollmentpolicy GetPoliciesResponse"`
// 	Response PolicyXCEPResponse `xml:"response"`
// 	CAs      string             `xml:"cAs"`
// 	OIDs     []XCEPoID          `xml:"oIDs"`
// }

// // PolicyXCEPResponse contains the policy information and its policies
// type PolicyXCEPResponse struct {
// 	PolicyID           string      `xml:"policyID"`
// 	PolicyFriendlyName string      `xml:"policyFriendlyName"`
// 	NextUpdateHours    interface{} `xml:"nextUpdateHours"`
// 	PoliciesNotChanged interface{} `xml:"policiesNotChanged"`
// 	Policies           XCEPPolicies
// }

// // XCEPPolicies contains the policies
// type XCEPPolicies struct {
// 	XMLName  xml.Name `xml:"policies"`
// 	Policies []XCEPPolicy
// }

// // XCEPPolicy contains the policies configuration
// type XCEPPolicy struct {
// 	XMLName        xml.Name       `xml:"policy"`
// 	OIDReferenceID int            `xml:"policyOIDReference"`
// 	CAs            interface{}    `xml:"cAs"`
// 	Attributes     XCEPAttributes `xml:"attributes"`
// }

// // XCEPAttributes contains the policies attributes
// type XCEPAttributes struct {
// 	PolicySchema              int                      `xml:"policySchema"`
// 	PrivateKeyAttributes      XCEPPrivateKeyAttributes `xml:"privateKeyAttributes"`
// 	SupersededPolicies        interface{}              `xml:"supersededPolicies"`
// 	PrivateKeyFlags           interface{}              `xml:"privateKeyFlags"`
// 	SubjectNameFlags          interface{}              `xml:"subjectNameFlags"`
// 	EnrollmentFlags           interface{}              `xml:"enrollmentFlags"`
// 	GeneralFlags              interface{}              `xml:"generalFlags"`
// 	HashAlgorithmOIDReference int                      `xml:"hashAlgorithmOIDReference"`
// 	RARequirements            interface{}              `xml:"rARequirements"`
// 	KeyArchivalAttributes     interface{}              `xml:"keyArchivalAttributes"`
// 	Extensions                interface{}              `xml:"extensions"`
// }

// // XCEPPrivateKeyAttributes contains attributes for the private key's generation and usage
// type XCEPPrivateKeyAttributes struct {
// 	MinimalKeyLength      int         `xml:"minimalKeyLength"`
// 	KeySpec               interface{} `xml:"keySpec"`
// 	KeyUsageProperty      interface{} `xml:"keyUsageProperty"`
// 	Permissions           interface{} `xml:"permissions"`
// 	AlgorithmOIDReference interface{} `xml:"algorithmOIDReference"`
// 	CryptoProviders       interface{} `xml:"cryptoProviders"`
// }

// // XCEPoID contains and OID value which can be referenced in a policy
// type XCEPoID struct {
// 	OIDReferenceID int    `xml:"policyOIDReference"`
// 	DefaultName    string `xml:"defaultName"`
// 	Group          int    `xml:"group"`
// 	Value          string `xml:"value"`
// }

// // NillableField can be used to specific no value (hence use the default)
// var NillableField = struct {
// 	Nil string `xml:"xsi:nil,attr"`
// }{
// 	Nil: "true",
// }
